pkg/volume/storageos/storageos.go

```diff
@@ -0,0 +1,786 @@
+}
+
+func (plugin *storageosPlugin) newUnmounterInternal(pvName string, podUID types.UID, manager storageosManager, mounter mount.Interface) (volume.Unmounter, error) {
+
+	// Parse volume namespace & name from mountpoint if mounted
+	volNamespace, volName, err := getVolumeInfo(pvName, podUID, plugin.host)
+	if err != nil {
+		return nil, err
+	}
+
+	return &storageosUnmounter{
+		storageos: &storageos{
+			podUID:          podUID,
+			pvName:          pvName,
+			volName:         volName,
+			volNamespace:    volNamespace,
+			manager:         manager,
+			mounter:         mounter,
+			plugin:          plugin,
+			MetricsProvider: volume.NewMetricsStatFS(getPath(podUID, volNamespace, volName, pvName, plugin.host)),
+		},
+	}, nil
+}
+
+func (plugin *storageosPlugin) NewDeleter(spec *volume.Spec) (volume.Deleter, error) {
+	if spec.PersistentVolume != nil && spec.PersistentVolume.Spec.StorageOS == nil {
+		return nil, fmt.Errorf("spec.PersistentVolumeSource.StorageOS is nil")
+	}
+
+	var secretNamespace, secretName string
+
+	// SecretName will be a locator ref for dynamically provisioned volumes
+	if strings.HasPrefix(spec.PersistentVolume.Spec.StorageOS.SecretName, secretLocatorPrefix) {
+		var err error
+		secretNamespace, secretName, err = plugin.lookupSecretRef(spec.PersistentVolume.Spec.StorageOS.SecretName)
+		if err != nil {
+			glog.Warningf("Could not locate API credentials for PV %q: %v", spec.PersistentVolume.Name, err)
+		}
+	}
+
+	// For pre-provisioned volumes there will not be a secretRef for the PV and
+	// although we have the secretName there's no way to get the secretNamespace.
+	// In most cases there will be a single secret in the default namespace, so
+	// using the default namespace at the risk that it is incorrect and the
+	// delete fails.  It is not possible to delete the wrong volume, the only
+	// risk is that the credentials retrieved do not have access.  If the secret
+	// is not in the default namespace the delete will fail.
+	if secretNamespace == "" {
+		secretNamespace = defaultNamespace
+	}
+	if secretName == "" {
+		secretName = spec.PersistentVolume.Spec.StorageOS.SecretName
+	}
+
+	manager, err := plugin.getManager(secretNamespace, secretName)
+	if err != nil {
+		return nil, err
+	}
+	return plugin.newDeleterInternal(spec, manager)
+}
+
+func (plugin *storageosPlugin) newDeleterInternal(spec *volume.Spec, manager storageosManager) (volume.Deleter, error) {
+
+	return &storageosDeleter{
+		storageos: &storageos{
+			pvName:       spec.Name(),
+			volName:      spec.PersistentVolume.Spec.StorageOS.VolumeName,
+			volNamespace: spec.PersistentVolume.Spec.StorageOS.VolumeNamespace,
+			manager:      manager,
+			plugin:       plugin,
+		},
+		pvUID: spec.PersistentVolume.UID,
+	}, nil
+}
+
+func (plugin *storageosPlugin) NewProvisioner(options volume.VolumeOptions) (volume.Provisioner, error) {
+	var secretName string
+	for k, v := range options.Parameters {
+		if strings.ToLower(k) == "secretname" {
```

Comment:
are these from a storageclass? if so, why not specify a secret namespace as a separate parameter? (c.f. https://github.com/kubernetes/kubernetes/blob/master/pkg/volume/quobyte/quobyte.go#L448)